flashlight/fl/tensor/backend/jit/ir/IndexNode.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "flashlight/fl/tensor/backend/jit/ir/IndexNode.h"
+
+#include <stdexcept>
+
+#include "flashlight/fl/tensor/backend/jit/JitTensorBase.h"
+
+namespace fl {
+
+namespace {
+
+Dim ceilDiv(Dim numerator, Dim denominator) {
+  return (numerator + denominator - 1) / denominator;
+}
+
+// TODO refactor with common index logic in OneDnnTensor; maybe put things like
+// this into a tensorBackendUtil file.
+Shape inferIndexedShape(const Shape& shape, const std::vector<Index>& indices) {
+  std::vector<Dim> indexedDims;
+  // TODO check indexing & shape validity, for now we let backend catch it
+  for (unsigned i = 0; i < indices.size(); i++) {
+    const auto& idx = indices[i];
+    switch (idx.type()) {
+      case detail::IndexType::Tensor: {
+        // TODO behavior of having more indices with the tensor index is a bit
+        // subtle, also uncommon. Rather than handling special cases, we don't
+        // support it at the moment.
+        if (indices.size() != 1) {
+          throw std::runtime_error(
+              "[inferIndexedShape]: Tensor must appear be the only index");
+        }
+        const auto& tensorIdx = idx.get<Tensor>();
+        // use tensor index's shape as base shape, and then append currently
+        // indexed-result shape onto the base shape (ignore the first dimension
+        // since it's reduced by tensor index)
+        //
+        // In Flashlight, the semantics is a bit different from numpy semantics:
+        // 1. if tensor index has different shape than indexed tensor, the
+        //    tensor index's is flattened first
+        // 2. else the result shape is just the tensor shape -- think `x(x < 0)`
```

Comment:
Can we also add a loud TODO here to adjust this behavior?

Would also be interested to see what PT does.

Reply:
Duplicated comment:)